Src/VimCore/CommonOperations.fs

```diff
@@ -149,6 +150,45 @@ type internal CommonOperations
         else
             _vimHost.Close _textView
 
+    /// Perform the specified action when the text view is ready
+    member x.DoActionWhenReady (action: unit -> unit) =
+
+        // Whether the text view is ready for the action
+        let isReady () = _vimHost.IsLoaded _textView
+
+        // Do the action if it is safe to do so. The window might have been
+        // closed in the meantime
+        let doAction () =
+            if not _textView.IsClosed then
+                action()
+
+        // Try to do the action if the text view is ready
+        let tryDoAction (bag: DisposableBag) =
+
+            // When it finally becomes ready, post the action because we should
+            // not perform any text view related actions from the text view's
+            // layout changed handler.
+            if isReady() then
+                let context = System.Threading.SynchronizationContext.Current
+                if context <> null then
+                    context.Post((fun _ -> doAction()), null)
+                else
+                    doAction()
+                bag.DisposeAll()
+
+        if isReady() then
+
+            // If the text view is ready, do the action immediately.
+            doAction()
+        else
+
+            // Otherwise, keep checking on each layout change whether the text
+            // view is ready for the action yet.
+            let bag = DisposableBag()
+            _textView.LayoutChanged
+            |> Observable.subscribe (fun _ -> tryDoAction bag)
+            |> bag.Add
+

```

Comment:
It occurred to me that we could convert `Vim.SetupInitialMode` to use this mechanism but I had a brain freeze trying to decide where it should go so that `Vim`, `CommonOperations`, and `SelectionChangeTracker` could all access it.

Reply:
It does seem like we should share the code between the this and `SetupInitialMode`; the coding pattern here is so similar that it seems to beg for a shared helper. Particularly because `SetupInitialMode` and this code has a different concept of "ready": one lets the host determine and the other checks for the presence of `ITextViewLines`. 

The code in `SelectionChangeTracker` is a bit different. It's not concerned with the "ready" state of the `ITextBuffer` nor does it use `LayoutChanged`. It only cares about the closed state. The pattern around `SynchronizationContext` is there though in all three places. 

After thinking about it a bit I believe there are two helpers we should add to `ICommonOperations`:

- `Post`:  this abstracts away the `SynchronizationContext.Post` pattern. The contract being the provided delegate will only run if the `ITextView` is still open after the post operation completes.
- `DoWhenReady`: the method you have today but delegating to `Post` for the `SynchronizationContext` bit. 







---
In reply to: [284557364](https://github.com/VsVim/VsVim/pull/2547#discussion_r284557364) [](ancestors = 284557364)

Reply:
All three cases want the code to run:

1. In the future but as soon as possible, and
2. Between when the text view becomes ready and when it is closed

We can and do receive an external caret position changed event before the window is ready for us to apply our view properties. It is a historical artifact that this code is even part of the selection change tracker as it has nothing directly to do with selection change tracking.

In the old code, the caret position changed event was only being watched in order to enforce scroll offset and so it had no effect when scroll offset was zero. It turns out that we need to enforce all our view properties (not just scroll offset) and so that event runs more often.

The problem with ICommonOperations is that IVim doesn't have direct access to ICommonOperations, maybe intentionally or maybe because of mutual dependency. If IVim even had the ICommonOperations factory, it could be done, but anyway that decision was what stalled my first attempt at unifying them.

Reply:
Changes based on your comments:

- I implemented a fully WPF-aware version of `DoActionWhenTextViewReady` in `IVimHost`
- I forwarded `DoActionWhenReady` in `ICommonOperations` to `DoActionWhenTextViewReady` in `IVimHost`
- I added `DoActionWhenReady` to the `ICommonOperations` interface
- I did the work necessary so that `IVim` can use `ICommonOperations` for a specific `IVimBuffer`
- I refactored `SetUpInitialMode` to use the `DoActionWhenReady`
- I moved `OnCaretPositionChanged` out of `SelectionChangeTracker` and into `Vim`